ccf_transaction_processor/scripts/stop_cchost.sh

```diff
@@ -23,6 +23,9 @@ fi
 F_SERVICEHOME="$( cd -P "$( dirname ${BASH_SOURCE[0]} )/.." && pwd )"
 source ${F_SERVICEHOME}/bin/lib/pdo_common.sh
 
-if [ -f ${F_SERVICEHOME}/run/cchost.pid ]; then
-    kill $(cat ${F_SERVICEHOME}/run/cchost.pid)
-fi
+pid_files=`find ${F_SERVICEHOME}/run -type f -name "cchost*.pid"`
+for file in ${pid_files}
+do
+    kill $(cat ${file})
+done
+
```

Comment:
this needs to be thought through more carefully. we do not need to test ccf by running multiple instances on a single server, it would be much better to assume multiple servers with one instance per (then use docker to run multiple on a server if necessary)

Reply:
Thank you for the question. My opinion is just the opposite. We do not need docker for deploying multiple nodes on the same server. Also multiple nodes on the same server ist just for testing and development, just like what we permit for pdo eservices. We want to use CCF docker only for easily deploying nodes on multiple machines. My expectation is that the current scripts should are all that we need even for docker as well.

Reply:
under what circumstances will we have more than one host on the server except for testing and if we are testing we'll be using docker. in this case, removing the test makes the application more unstable. that is, under normal circumstances you don't want a separate application running (for start) or make sure its running (stop)

Reply:
Thank you, you are right. more than one node on a single node is only for the development phase. Given that CCF is several months away from stable 1.0, (they are already talking about .12 with breaking apis), I would like to have the ability to local test multiple nodes with just one VM. and I am not in favor of using docker for code development and local testing (much like what we do for pdo) for the multi-node setup. For multiple machines, using docker is my preference., and I suggest that that be our only supported deployment model.

Reply:
agreed and this is what docker is for so we should not complicate the deployment and configuration of the current code. I would strongly suggest that we drop the notion of multiple instances on a single server and focus on accelerating th docker work. that's a much better solution.

Reply:
I plan to create the dockerfiles around the multi-node scripts that we have in this PR. The multinode scripts have been written in my mind that permits easy docker file creation, once this PR is approved. I am going to push a commit to this PR addressing some of your review concerns. The contents of this PR will be there even in the docker PR, so just thought to separate into two.  I am afraid, I dont think we can drop the notion of multiple instances on single server and we need to carry this PR in our repo until CCF becomes stable, since the APIs for multi-node deployment also keeps changing (not just the CCF APP apis). For instance, if you see here, there are several changes with repsect to 0.9 in the node deployment.  So even if we have docker for a real-life deployment, testing the CCF's multi-node APIs needs a set up like what we have in this PR before can go with docker containers. Yes, the docker PR is coming soon

Reply:
can you help me understand why we would ever need multiple instances on a server? remember... CCF is NOT our project. we do not need to test CCF. the only circumstances where we would ever need to test a multinode CCF is if we are concerned about performance implications. otherwise multinode CCF will only occur in customer deployments and the tests performed there will be CCF tests, not PDO tests. Dockerizing a single instance seems to be, by far, the highest priority here because we cannot replace sawtooth in the testing until that is done. maybe we should split this PR into two parts, one that upgrades CCF and a separate one that will not be pushed until the dockerization is complete. then all multi-node testing would be done through the tools that docker provides.

Reply:
I am not saying that we have to test CCF. My limited point is that as long as we offer multi-node support for deploying PDO TP via CCF, even if it is via docker as you say, there will be changes that we need to make to the multi-node deployment scripts (if it is docker, scripts that get referenced within dockerfiles) because of the breaking (deployment-related) API changes with newer CCF releases. I do not want to complicate fixing this API by adding a docker layer on top. I would imagine that we write dockerfiles in such a way that the dockerfiles itself remain stable, and only the scripts referenced within the dockerfiles that need change because of CCF folks making breaking changes. These scripts are exactly the ones that are introduced int this PR. It will be much easier to ensure that the scripts work without the added complexity docker, and once the scripts are fixed, docker should also automatically work.

Reply:
multi-node deployment is a function of CCF & should use their tools, not ours. this is the same approach we had with sawtooth. the only thing we need for multinode is enough support to run experiments for a paper.

Reply:
and... the point i made earlier is that, for example, you removed the pid management & testing from the script that i would expect everyone to use to start the CCF validator. so you are essentially forcing me to use the multinode starter even though i'm effectively NEVER going to use it.

Reply:
1. We are using CCF's tools such as cchost & CCFClient. all that we are doing is to wrap these around scripts that can automate the set up. 
2. I don't agree that we are forcing anyone to use to a multi-node starter. By default, the start_ccf_network will start a single node only, and the operator does not even need to set any input for this to happen. To the extent possible, we have separated scripts that handle multiple nodes including the tomls files, the add_node script. It is only the stop node script that tries to stop all running nodes, and it is only this script that is resued for both single node and multi node set up. I can wrap this also around another mult-node-stop script that will be used only in the case of  multi-node and this way the operator does not see any multi-node script in their pipeline.

Reply:
There is one more important reason why I strongly think we need to be able to locally test CCF multiple nodes. When CCF is operated with multiple nodes, the read rpcs can go to any of the nodes, and in a true deployment, this can potentially remove the bottleneck of the RAFT head. (given that pdo execution involves several reads in one invocation). This mechanism is partially enforced via the PDO TP itself (by installing certain rpcs as reads, and returning only globally committed values), and we need to consistently make sure that the software works (both PDO-client and PDO-TP) to address potential changes in both the PDO side as well CCF side (both CCF version changes as well any changes to TP). I do not want a situation where the functionality aspects of the PFO/mullti-node CCF cannot be tested without deploying docker.

Reply:
we aren't using their tools... you wrote your own script. anyone who is deploying CCF in anything like a "production" environment is going to use deployment scripts from Azure or Docker, not our toy scripts. and your script is inconsistent with any deployment we (PDO developers) ever expect to use for CCF. we need the multinode in PDO because we need to test the protocols for multiple instances (e.g. synchronization of state). we do not need to test any multi-node aspects of CCF (they can handle their own consensus tests). again... i'm having a hard time understanding any situation where start_ccf_network would be useful (beyond what we already had in start_cchost). could you give me an example where we would benefit from a multi-node setup? (and performance testing is part of paper development, not PDO development)

Reply:
I hope my last comment on being able to route reads to any CCF node answers your question as to where we would benefit from a multi-node set up on the same server. As a developer, I would need the ability to verify functionality on a local VM of the whole system rather than having to deploy multiple machines or dockers.

Reply:
There might have been a slight misunderstanding between us. All along whenever I said testing, I meant "functionality verification" and not "performance benchmarking"

Reply:
Besides, I thought we were also discussing the possibility of introducing "per ccf node" ledger authority keys rather than one global key to make the CCF ledger authority more robust. This just like the "reads from any ccf node" is a protocol aspect that will be implemented via the PDO TP and PDO client, and we need easy mechanisms to verify the functionality of PDO/multi-node-ccf.